src/emulator/dataconnectEmulator.ts

```diff
@@ -97,12 +98,31 @@ export class DataConnectEmulator implements EmulatorInstance {
           `FIREBASE_DATACONNECT_POSTGRESQL_STRING is set to ${dataConnectLocalConnString()} - using that instead of starting a new database`,
         );
       } else {
-        connStr = `postgres://${this.args.postgresHost ?? "127.0.0.1"}:${this.args.postgresPort ?? 5432}/${dbId}?sslmode=disable`;
         const pgServer = new PostgresServer(dbId, "postgres");
-        const server = await pgServer.createPGServer(
-          this.args.postgresHost,
-          this.args.postgresPort,
-        );
+        if (this.args.postgresPort) {
+          const process = await lsofi(this.args.postgresPort);
+          if (process) {
+            const errMessage =
+              `Data Connect: Unable to start PGLite server on port ${this.args.postgresPort} because it is already in use by process number ${process}. This may occur if you are running another instance of Postgres.` +
+              ` You can choose a different port by setting 'firebase.json#emulators.dataconnect.postgresPort', or you can unset that field to automatically find an open port.`;
+            throw new FirebaseError(errMessage);
+          }
+        }
+        const port = this.args.postgresPort || (await findOpenPort(5432));
```

Comment:
This should follow Firestore emulator websocket's example to find a port early and avoid conflicts: https://github.com/firebase/firebase-tools/blob/0314586ffffcfe286caacb067465c1817286c806/src/emulator/controller.ts#L398

I'm approving to unblock but please clean this up as a follow up. Please also remove `findOpenPort` -- it only checks `127.0.0.1`, not `postgresHost`, while the `resolveHostAndAssignPorts` as linked above handles all these edge cases

Reply:
Addressed in #7727 (no pun intended)